R/computeAnnualDegreeDays.R

```diff
@@ -0,0 +1,191 @@
+                       "tlim"     = t)    # [C]
+
+              return(hddcddAgg)
+            }
+          )
+        )
+      }
+    )
+  )
+
+  return(hddcdd)
+}
+
+
+
+#' Assign HDD/CDD values for given ambient/limit temperature
+#'
+#' @param temp SpatRaster containing temperature/BAIT values
+#' @param typeDD type of degree day
+#' @param tlim limit temperature
+#' @param factors data frame with degree day values for \code{temp/tlim} combination
+#'
+#' @return SpatRaster with HDD/CDD values
+#'
+#' @author Hagen Tockhorn
+#'
+#' @importFrom terra classify tapp time
+#' @importFrom dplyr filter reframe .data
+#' @importFrom magrittr %>%
+
+compCellHDDCDD <- function(temp, typeDD, tlim, factors) {
+  # extract years
+  dates <- names(temp)
+
+  # add tolerance of 0.049K to avoid machine precision errors
+  factors <- factors[factors$typeDD == typeDD, ]
+
+  factors <- factors %>%
+    filter(.data[["tLim"]] == tlim) %>%
+    dplyr::reframe(from =    .data[["T_amb_K"]] - 0.049,
+                   to =      .data[["T_amb_K"]] + 0.049,
+                   becomes = .data[["factor"]]) %>%
+    data.matrix()
+
+  # swap ambient temperature values with corresponding DD values
+  hddcdd <- classify(temp, factors)
+
+  terra::time(hddcdd) <- as.Date(dates)
+
+  # aggregate to yearly HDD/CDD [K.d/a]
+  hddcdd <- tapp(hddcdd, "years", fun = sum, na.rm = TRUE)
+
+  names(hddcdd) <- gsub("y_", "", names(hddcdd))
+  return(hddcdd)
+}
+
+
+
+#' Aggregate cellular HDD/CDD values to country-wide average (population-weighted)
+#'
+#' @param data SpatRaster object containing HDD/CDD values
+#' @param weight SpatRaster object containing aggregation weights
+#' @param mask SpatRaster object defining (regional) aggregation boundaries
+#'
+#' @return data frame containing regionally averaged HDD/CDD values
+#'
+#' @author Hagen Tockhorn
+#'
+#' @importFrom terra subset global
+
+aggCells <- function(data, weight, mask) {
+  message("Aggregating degree days to regions...")
+
+  yearsData   <- names(data)
+  yearsWeight <- names(weight)
+
+  if (!all(yearsData %in% yearsWeight)) {
+    stop("Time periods of raster file and aggregation weights do not match.")
+  }
+
+  # loop: years in raster file r
+  hddcddAgg <- do.call(
+    "rbind", lapply(
+      yearsData, function(y) {
+        # mask data and weights to considered regions
+        regData   <- subset(data, y) * subset(weight, y) * mask
+        regWeight <- subset(weight, y) * mask
+
+        # aggregate regional data
+        regDataAgg   <- terra::global(regData,   "sum", na.rm = TRUE)$sum
+        regWeightAgg <- terra::global(regWeight, "sum", na.rm = TRUE)$sum
+
+        # calculate weighted sum
+        weightedAgg <- regDataAgg / regWeightAgg
+
+        tmp <- data.frame("region" = names(mask),
+                          "period" = y,
+                          "value"  = round(weightedAgg, 3))
+
+        rm(regData, regWeight, regDataAgg, regWeightAgg, weightedAgg)
+
+        rownames(tmp) <- c()
```

Comment:
I'm surprised and expected rather something like this to remove row names:
```suggestion
        rownames(tmp) <- NULL
```